#!/usr/bin/env python
# vim: set syntax=python :

PACKAGE = "ScaViSLAM"
from dynamic_reconfigure.parameter_generator_catkin import *

gen = ParameterGenerator()

gen.add( "wordspath", str_t, 0,
        "Path to place finder words file", "data/surfwords10000.png")

gen.add( "queue_size", int_t, 0,
        "Image synchronization queue size", 5, 0, 20)

gen.add( "approximate_sync", bool_t, 0,
        "Use approximate synchronization on inputs", True)

gen.add( "parallax_threshold", double_t, 0,
         "Point motion threshold to add keyframe", 0.75, 0, 2)

gen.add( "newpoint_clearance", int_t, 0,
         "Pixel distance between keypoints", 2, 0, 50)

gen.add( "covis_threshold", int_t, 0,
         "Minimum number of shared points to connect two keyframes", 15, 5, 50)

gen.add( "new_keyframe_featureless_corners_thr", int_t, 0,
         "Featureless corners to trigger new keyframe addition", 2, 0, 4)

gen.add( "use_n_levels_in_frontend", int_t, 0,
         "Pyramid levels for feature matching", 2, 1, 4)

gen.add( "ui_min_num_points", int_t, 0,
         "Any 3x3 window must have at least this many points for newpoints to be added", 25, 20, 200)

gen.add( "num_disp16", int_t, 0,
         "Number of stereo disparities", 2, 1, 10 )

stereo_method_enum = gen.enum([
    gen.const("BM", int_t, 1, "Block Matching"),
    gen.const("GPU_BM", int_t, 2, "GPU accelerated block matching"),
    gen.const("GPU_BP", int_t, 3, "GPU belief propagation"),
    gen.const("GPU_CSBP", int_t, 4, "GPU constant space belief propagation")],
    "Stereo Correspondence calculation method")
gen.add( "stereo_method", int_t, 0,
         "Stereo Correspondence calculation method", 2, 1, 4,
         edit_method=stereo_method_enum)

gen.add( "stereo_iters", int_t, 0,
         "Stereo CSBP iterations", 4, 1, 20)

gen.add( "stereo_levels", int_t, 0,
         "Stereo CSBP levels", 4, 1, 5)

gen.add( "stereo_nr_plane", int_t, 0,
         "Stereo CSBP levels", 1, 1, 5)

gen.add( "var_num_max_points", int_t, 0,
         "Stop adding points after this many new points found", 300, 50, 1000)

gen.add( "max_reproj_error", double_t, 0,
         "Maximum reprojection error to use a matched point", 2, 0, 5)

exit(gen.generate(PACKAGE, "stereo_vslam_node", "StereoVSLAM"))
